src/omv/imlib/simd.h

```diff
@@ -561,6 +596,90 @@ static inline v128_t vusat_s16_narrow_u8_hi(v128_t v0, v128_t v1, uint32_t shift
 }
 #endif
 
+// There's a software implementation of __USAT in the ARM CMSIS extension if needed
+#define usat_s16(x, bits) __USAT(x, bits)
+
+// There's a software implementation of __SSAT in the ARM CMSIS extension if needed
+#define ssat_s16(x, bits) __SSAT(x, bits)
+
+#if (__ARM_ARCH >= 8)
+#define vusat_s16_narrow_u8(v0, bits) \
+    ((v128_t) vshrq_n_u16(vqshluq_n_s16(v0.s16, 16 - bits), 16 - bits))
+#else
+static inline v128_t vusat_s16_narrow_u8(v128_t v0, uint32_t bits) {
+    // There's a software implementation of __USAT16 in the ARM CMSIS extension if needed
+    return (v128_t) {
+        .u32 = { __USAT16(v0.u32[0], bits) }
+    };
+}
+#endif
+
+#if (__ARM_ARCH >= 8)
+#define vusat_s16_narrow_u8_combine(v0, v1) \
+    ((v128_t) vqmovuntq_s16(vqmovunbq_s16(vuninitializedq_s16(), v0.s16), v1.s16))
+#else
+static inline v128_t vusat_s16_narrow_u8_combine(v128_t v0, v128_t v1) {
+    // There's a software implementation of __USAT16 in the ARM CMSIS extension if needed
+    return (v128_t) {
+        .u32 = { __USAT16(v0.u32[0], 8) | (__USAT16(v1.u32[0], 8) << 8) }
+    };
+}
+#endif
+
+#if (__ARM_ARCH >= 8)
+#define vmov_clean_u16_narrow_u8_combine(v0, v1) ((v128_t) vmovntq(v0.u8, v1.u16))
+#else
+static inline v128_t vmov_clean_u16_narrow_u8_combine(v128_t v0, v128_t v1) {
+    return (v128_t) {
+        .u32 = { v0.u32[0] | (v1.u32[0] << 8) }
+    };
+}
+#endif
+
+#if (__ARM_ARCH >= 8)
+#define vmov_dirty_u16_narrow_u8_combine(v0, v1) ((v128_t) vmovntq(v0.u8, v1.u16))
+#else
+static inline v128_t vmov_dirty_u16_narrow_u8_combine(v128_t v0, v128_t v1) {
```

Comment:
No, I don't want a mask at all for the clean version since I don't need the AND operation. The clean version is meant for narrowing 16-bit to 8-bit values where I know the top 8-bits of the 16-bits are already 0. With Helium, it always masks the top 8-bits and drops them. So, it doesn't matter. On non-Helium I'd have an extra AND with a mask added.

Reply:
The clean version is used for stuff like this:

```
// Divide by ksize.
v128_t pixels0 = vdiv_fast_s16(vacc0, vdiv_ksize);
v128_t pixels1 = vdiv_fast_s16(vacc1, vdiv_ksize);
v128_t pixels = vmov_clean_u16_narrow_u8_combine(pixels0, pixels1);
```

After doing the fast div here ((value * mult) >> 16) I know the result will be an 8-bit number in each 16-bit slot. So, you just need a shift and then or which compiles down to one instruction in ARM non-Helium.

Reply:
Conversely, the dirty version is needed for stuff like:

vrgb_pixels888_to_grayscale

Which takes 8-bit values occupying 16-bits, and then generates a 16-bit values where the bottom 8-bits are the grayscale value and the top 8-bits of the 16-bits are garbage. So, you have to mask out the bottom bits to combine things.

Reply:
> On non-Helium I'd have an extra AND with a mask added.

Yes, but older architectures are slow anyway, extra and will remove duplicate code, so it could help with text size. For Helium, and newer, the arg will be unused.

Reply:
Combined them into one with an arg to turn the mask on.